src/FluffySpoon.AspNet.LetsEncrypt.Aws/AwsDnsChallengePersistenceStrategy.cs

```diff
@@ -0,0 +1,204 @@
+
+			var recordSet = new ResourceRecordSet
+			{
+				Name = recordName,
+				TTL = 60,
+				Type = RRType.FindValue(recordType),
+				ResourceRecords = new List<ResourceRecord> { new ResourceRecord { Value = recordValue } }
+			};
+
+			var change1 = new Change
+			{
+				ResourceRecordSet = recordSet,
+				Action = ChangeAction.UPSERT
+			};
+
+			var changeBatch = new ChangeBatch
+			{
+				Changes = new List<Change> { change1 }
+			};
+
+			var recordsetRequest = new ChangeResourceRecordSetsRequest
+			{
+				HostedZoneId = zone.Id,
+				ChangeBatch = changeBatch
+			};
+
+			_logger.LogInformation($"Creating or updating DNS record {recordType} {recordName}");
+
+			var upsertResponse = await _route53Client.ChangeResourceRecordSetsAsync(new ChangeResourceRecordSetsRequest() { ChangeBatch = changeBatch, HostedZoneId = zone.Id });
+
+			var changeRequest = new GetChangeRequest
+			{
+				Id = upsertResponse.ChangeInfo.Id
+			};
+
+			while ((await _route53Client.GetChangeAsync(changeRequest)).ChangeInfo.Status == ChangeStatus.PENDING)
+			{
+				_logger.LogDebug("Creation/update of {RecordType} {RecordName} with value {RecordValue} is pending. Checking for status update in {StatusPollIntervalSeconds} seconds.", recordType, recordName, recordValue, StatusPollIntervalSeconds);
+				Thread.Sleep(TimeSpan.FromSeconds(StatusPollIntervalSeconds));
+			}
+		}
+
+		private async Task<HostedZone> FindHostedZone(string dnsName)
+		{
+			_logger.LogDebug("Finding hosted zone responsible for {DnsName}", dnsName);
+
+			var partsToMatch = dnsName.Split(DomainSegmentSeparator);
+			var isWildcard = dnsName.StartsWith(WildcardPrefix);
+			var bestPossibleScore = isWildcard ? partsToMatch.Length - 1 : partsToMatch.Length;
+
+			HostedZone bestMatch = null;
+			int bestMatchScore = 0;
+
+			// Enumerate all hosted zones until a match is found, or until all hosted zones have been enumerated
+			var hostedZones = await _route53Client.ListHostedZonesAsync();
+			while (hostedZones.IsTruncated)
+			{
+				foreach (var hostedZone in hostedZones.HostedZones)
+				{
+					_logger.LogDebug("Checking zone {Zone}", hostedZone.Name);
+
+					if (dnsName.ToLower().EndsWith(DomainSegmentSeparator + hostedZone.Name.ToLower()))
+					{
+						// Matches, calculate score
+						var hostedZoneParts = hostedZone.Name.Split(DomainSegmentSeparator);
+						var score = hostedZoneParts.Length;
+
+						if (score == bestPossibleScore)
+						{
+							_logger.LogInformation("Exact match for {DnsName} found (zone {Zone})", dnsName, hostedZone.Name);
+
+							// Exact match found
+							return hostedZone;
+						}
+						else if (score > bestMatchScore)
+						{
+							_logger.LogDebug("Setting best match for {DnsName} to zone {Zone}", dnsName, hostedZone.Name);
+
+							bestMatch = hostedZone;
+							bestMatchScore = score;
+						}
+					}
+				}
+
+				hostedZones = await _route53Client.ListHostedZonesAsync(new ListHostedZonesRequest() { Marker = hostedZones.Marker });
+			}
```

Comment:
Extracting that into a function would just result in another function of practically the same size serving the same purpose, only ever used in one place. I don't see that being of any benefit.

Reply:
I've reviewed it a few times, and because of the interaction between variables, I've not come up with a way to simplify/refactor it in any constructive way.